File: src/sys/dotnet/fan/sys/SysInStream.cs
//
// Copyright (c) 2007, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   12 Mar 07  Andy Frank  Creation
//

using System;
using System.IO;
using System.Text;

namespace Fan.Sys
{
  /// <summary>
  /// SysInStream.
  /// </summary>
  public class SysInStream : InStream
  {

  //////////////////////////////////////////////////////////////////////////
  // Constructor
  //////////////////////////////////////////////////////////////////////////

    public static SysInStream make(Stream input, Long bufSize)
    {
      if (bufSize == null || bufSize.longValue() == 0)
        return new SysInStream(input);
      else
        return new SysInStream(new BufferedStream(input, bufSize.intValue()));
    }

    public SysInStream(Stream input)
    {
      this.inStream = input;
    }

  //////////////////////////////////////////////////////////////////////////
  // Obj
  //////////////////////////////////////////////////////////////////////////

    public override Type @typeof() { return Sys.SysInStreamType; }

  //////////////////////////////////////////////////////////////////////////
  // InStream
  //////////////////////////////////////////////////////////////////////////

    public override Long read() { int n = r(); return n < 0 ? null : Long.valueOf(n); }
    public override int r()
    {
      try
      {
        return inStream.ReadByte();
      }
      catch (IOException e)
      {
        throw IOErr.make(e).val;
      }
    }

    public override Long readBuf(Buf buf, long n)
    {
      try
      {
        long read = buf.pipeFrom(inStream, n);
        if (read <= 0) return null;
        return Long.valueOf(read);
      }
      catch (IOException e)
      {
        throw IOErr.make(e).val;
      }
    }

    public override InStream unread(long n) { return unread((int)n); }
    public override InStream unread(int n)
    {
      try
      {
        // don't take the hit until we know we need to wrap
        // the raw input stream with a pushback stream
        if (!(inStream is PushbackStream))
          inStream = new PushbackStream(inStream, 128);
        ((PushbackStream)inStream).Unread(n);
        return this;
      }
      catch (IOException e)
      {
        throw IOErr.make(e).val;
      }
    }

    public override long skip(long n)
    {
      try
      {
        for (int i=0; i<n; ++i)
          if (r() < 0) return i;
        return n;
      }
      catch (IOException e)
      {
        throw IOErr.make(e).val;
      }
    }

    public override bool close()
    {
      try
      {
        if (inStream != null) inStream.Close();
        return true;
      }
      catch (IOException)
      {
        return false;
      }
    }

  //////////////////////////////////////////////////////////////////////////
  // PushbackStream
  //////////////////////////////////////////////////////////////////////////

    internal class PushbackStream : Stream
    {
      public PushbackStream(Stream parent, int size)
      {
        this.parent = parent;
        this.pushback = new byte[size];
      }

      //
      // Properties
      //

      public override bool CanRead  { get { return parent.CanRead;  } }
      public override bool CanWrite { get { return parent.CanWrite; } }
      public override bool CanSeek  { get { return false; } } // don't allow seeking
      public override long Length   { get { return parent.Length; } }
      public override long Position
      {
        get { return parent.Position;  }
        set { parent.Position = value; }
      }

      //
      // Methods
      //

      public void Unread(int b)
      {
        if (pos >= pushback.Length)
          throw new IOException("Not enough room in pushback buffer");
        pushback[pos++] = (byte)b;
      }

      public override int Read(byte[] buf, int offset, int count)
      {
        // check pushback first
        if (pos > 0)
        {
          int len = Math.Min(pos, count);
          Array.Copy(pushback, pos-len, buf, offset, len);
          pos -= len;
          offset += len;
          count -= len;

          // short-circuit if buf filled with pushback
          if (count == 0) return len;
        }

        // grab remaining bytes
        return parent.Read(buf, offset, count);
      }

      public override void Write(byte[] buf, int offset, int count)
      {
        // This is purely an input stream, so Write not allowed
        throw new IOException("PushbackStream does not allow Write()");
      }

      public override long Seek(long offset, SeekOrigin origin)
      {
        // We don't allow seeking
        throw new IOException("PushbackStream does not allow Seek()");
      }

      public override void SetLength(long val)
      {
        // We don't allow seeking
        throw new IOException("PushbackStream does not allow SetLength()");
      }

      public override void Close() { parent.Close(); }
      public override void Flush() { parent.Flush(); }

      //
      // Fields
      //

      private Stream parent;
      private byte[] pushback;
      private int pos = 0;
    }

  //////////////////////////////////////////////////////////////////////////
  // .NET Conversion
  //////////////////////////////////////////////////////////////////////////

    /// <summary>
    /// Get a System.IO.Stream for the specified input stream.
    /// </summary>
    public static Stream dotnet(InStream ins)
    {
      if (ins is SysInStream)
        return ((SysInStream)ins).inStream;
      else
        return new DotnetInputStream (ins);
    }

    internal class DotnetInputStream : Stream
    {
      public DotnetInputStream(InStream ins) { this.ins = ins; }

      // Properties
      public override bool CanRead  { get { return true;  } }
      public override bool CanWrite { get { return false; } }
      public override bool CanSeek  { get { return false; } }
      public override long Length   { get { return -1; } }
      public override long Position { get { return -1; } set {} }

      // Methods
      public override int Read(byte[] b, int off, int len)
      {
        buf.m_buf = b;
        buf.m_pos = off;
        buf.m_size = b.Length;
        Long n = ins.readBuf(buf, len);
        buf.m_buf = null;
        if (n == null) return -1;
        return n.intValue();
      }
      public override long Seek(long off, SeekOrigin origin) { return -1; }
      public override void SetLength(long val) {}
      public override void Flush() {}
      public override void Write(byte[] b, int off, int len) {}
      public override void Close() { ins.close(); }

      // Fields
      private InStream ins;
      private MemBuf buf = new MemBuf(null, 0);
    }

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    internal Stream inStream;

  }
}

File: src/sys/java/fan/sys/SysInStream.java
//
// Copyright (c) 2006, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   27 Mar 06  Brian Frank  Creation
//
package fan.sys;

import java.io.*;

/**
 * SysInStream implements InStream using a java.io.InputStream
 */
public class SysInStream
  extends InStream
{

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

  public static SysInStream make(java.io.InputStream in, Long bufSize)
  {
    if (bufSize == null || bufSize.longValue() == 0)
      return new SysInStream(in);
    else
      return new SysInStream(new java.io.BufferedInputStream(in, bufSize.intValue()));
  }

  public SysInStream(InputStream in)
  {
    this.in = in;
  }

//////////////////////////////////////////////////////////////////////////
// Obj
//////////////////////////////////////////////////////////////////////////

  public Type typeof() { return Sys.SysInStreamType; }

//////////////////////////////////////////////////////////////////////////
// InStream
//////////////////////////////////////////////////////////////////////////

  public long avail()
  {
    try
    {
      return in.available();
    }
    catch (IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public Long read() { int n = r(); return n < 0 ? null : FanInt.pos[n]; }
  public int r()
  {
    try
    {
      return in.read();
    }
    catch (IOException e)
    {
      throw IOErr.make(e);
    }
  }

  /**
   * Bulk read the remaining bytes and decode them in one pass
   */
  public String readAllStr(boolean normalizeNewlines)
  {
    try
    {
      MemBuf buf = readAllBytes();
      String s = charsetDecoder.decodeAll(buf.buf, buf.size);
      if (normalizeNewlines && s.indexOf('\r') >= 0) s = s.replace("\r\n", "\n").replace('\r', '\n');
      return s;
    }
    finally
    {
      try { close(); } catch (Exception e) { e.printStackTrace(); }
    }
  }

  /**
   * Read remaining bytes, sized by available() so a file fits in one read
   */
  private MemBuf readAllBytes()
  {
    try
    {
      byte[] buf = new byte[Math.max(in.available() + 1, 1024)];
      int size = 0;
      while (true)
      {
        if (size == buf.length) buf = java.util.Arrays.copyOf(buf, size * 2);
        int n = in.read(buf, size, buf.length - size);
        if (n < 0) break;
        size += n;
      }
      return new MemBuf(buf, size);
    }
    catch (IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public Long readBuf(Buf buf, long n)
  {
    try
    {
      long read = buf.pipeFrom(in, n);
      if (read < 0) return null;
      return Long.valueOf(read);
    }
    catch (IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public InStream unread(long n) { return unread((int)n); }
  public InStream unread(int n)
  {
    try
    {
      // don't take the hit until we know we need to wrap
      // the raw input stream with a pushback stream
      if (!(in instanceof PushbackInputStream))
        in = new PushbackInputStream(in, 128);
      ((PushbackInputStream)in).unread(n);
      return this;
    }
    catch (IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public long skip(long n)
  {
    try
    {
      long skipped = 0;
      while (skipped < n)
      {
        long x = in.skip(n-skipped);
        if (x < 0) break;
        skipped += x;
      }
      return skipped;
    }
    catch (IOException e)
    {
      throw IOErr.make(e);
    }
  }

  public boolean close()
  {
    try
    {
      if (in != null) in.close();
      return true;
    }
    catch (IOException e)
    {
      return false;
    }
  }

//////////////////////////////////////////////////////////////////////////
// Java Conversion
//////////////////////////////////////////////////////////////////////////

  /**
   * Get a java.io.InputStream for the specified input stream.
   */
  public static InputStream java(InStream in)
  {
    if (in instanceof SysInStream)
      return ((SysInStream)in).in;
    else
      return new JavaInputStream(in);
  }

  static class JavaInputStream extends InputStream
  {
    JavaInputStream(InStream in) { this.in = in; }

    public int read()
    {
      return in.r();
    }

    public int read(byte[] b, int off, int len)
    {
      buf.buf = b;
      buf.pos = off;
      buf.size = b.length;
      Long n = in.readBuf(buf, len);
      buf.buf = null;
      if (n == null) return -1;
      return n.intValue();
    }

    public void close() { in.close(); }

    InStream in;
    MemBuf buf = new MemBuf(null, 0);
  }

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  InputStream in;

}

